### PR TITLE
Add command RemoveOldReports

### PR DESCRIPTION
**Pull request by DanielMurphy22**
- A new command has been added to "Remove Old Reports" actually it removes the user identifiable info from the reports and deletes UserDetails that are not tied to and Active report (one that hasn't had it's user info removed, as it occured in the last 90 days). The `id` attached to old ErrorReport hasn't been deleted, but I believe this corresponds to the name and email of UserDetails which is getting deleted!
- This was based off of a [command file to remove old recovery files](https://github.com/mantidproject/errorreports/blob/master/web/services/management/commands/removeolddata.py) (which I have replaced as removing recovery files is useless now we no longer accept recovery files!)
- Also, when I launched docker there was a migration file (from another PR: https://github.com/mantidproject/errorreports/commit/38584f03ea1606f55dcfec87e2e9efb3bd4ef621) which had not been logged in this repo, so I've included it here.

@martyngigg  I assume a Cron job can be run on the actual server along the lines of 
`0 0 * * * docker exec errorreports_web_1 python manage.py removeoldreports`

To test:

On a local server (with mantid.properties: errorreports.rooturl = http://localhost:8082).
Then create some error reports! I created about 5 reports, waited 3 minutes and created more. 
You will need to edit the command file (web/services/management/commands/removeoldreports.py) to filter for 5 minutes rather than 90 days. To do this, change line 26 and 48 to: `...timedelta(**minutes=options['days'])**)`

To actually delete user info from the old reports, (assuming the server is already up!) you need to run: 
`docker exec errorreports_web_1 python manage.py removeoldreports`
This defaults to run the "days" command (which would usually delete reports older than 90 days, but you've set this to 90 mins!), if you add the `--all` argument, then as expected, all error reports will be deleted. 

You probably want to run: `docker exec errorreports_web_1 python manage.py removeoldreports 5` to delete reports older than 5 mins.
By the time you have shutdown and re-booted the local server, the first handful of reports will be <5 mins old and the later ones will still be in the last 5 mins! Good luck with your timings!

To create error reports: load `TrainingCourseData/Training_Exercise3a_SNS.nxs`, executing `NormaliseToMonitor` algorithm and clicking "Yes, Share Info"
Repeat as necessary!

### CODE REVIEW

**Comment by DanielMurphy22** [comment]
@Mantid-Matthew I think it would make sense for David to review this, when he get a chance, but it would be great for you to have a look through as well. 
In particular, I've updated the README a little so tell me if that makes sense!

**Comment by DanielMurphy22** [comment]
What I haven't added in here is removing the UserDetails that have no attributed reports!

**Comment by DanielMurphy22** [comment]
Something I haven't done, but @NickDraper mentioned breifly was that in deleting reports we lose stats on ErrorReports, e.g. How many per month, how many from each facility, how many from different OSs? 

We could save out some of this info before deleting, or just be happy with losing this info? I won't get to this Today, so I'll leave to @DavidFair and @Mantid-Matthew to consider!

**Comment by Mantid-Matthew** [comment]
> Something I haven't done, but @NickDraper mentioned breifly was that in deleting reports we lose stats on ErrorReports, e.g. How many per month, how many from each facility, how many from different OSs?
> 
> We could save out some of this info before deleting, or just be happy with losing this info? I won't get to this Today, so I'll leave to @DavidFair and @Mantid-Matthew to consider!

From a GDPR perspective if we remove the stacktrace, hashed user information, remove the link to the user table and then clear out old user information not linked to any report, we should be fine to keep the older reports. So we can remove all the identifying information and still keep the information useful from a statistical perspective.

File: web/services/management/commands/removeoldreports.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        # Delete identifiable parts of chosen reports
+        reports.update(uid="")
+        reports.update(host="")
+        reports.update(user_id="")
+        reports.update(textBox="")
+        reports.update(stacktrace="")
+
+        """Get UserDetails which are not tied to any surviving ErrorReports"""
+
+        users_with_surviving_reports = ErrorReport.objects.exclude(
```
Comment (DavidFair) [inline web/services/management/commands/removeoldreports.py:48]: ... call the `UserDetailsModel` to `removeOrphaned()` with something similar to [this SO answer](https://stackoverflow.com/a/6205303)

This gets SQL to do all the heavy lifting, which it was designed for ;)
```python
from django.db.models import Count

class UserDetails(models.Model):
    def clearOrphanedRecords():
        # First sort all UserDetailRecords by their F.K. count "user`, then using a named pipe
        # Filter out all records equal to 0
        no_refs = UserDetails.objects.annotate(num_refs=Count('user')).filter(num_refs__eq=0)
        no_refs.delete()
```

File: web/services/management/commands/removeoldreports.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        # with NO active/surviving error reports
+        for index in range(len(all_user)):
+            (User_ID,) = all_user.values_list('id')[index]
+            all_user_list.append(User_ID)
+            if all_user_list[index] not in active_users:
+                empty_users_objects.append(all_user[index])
+                empty_users.append(User_ID)
+
+        # Print User IDs for deleting
+        print("Active Users:")
```
Comment (DavidFair) [inline web/services/management/commands/removeoldreports.py:75]: We shouldn't print this as were moving PII into the logs unintentionally, which seldom get rotated

File: README.md [hunk 3]
```diff
@@ -36,21 +42,22 @@ To start the services locally you will first need to create a `.env` file next t
...
+        EMAIL_FROM_ADDRESS=error-reports@mantidproject.org
+        SLACK_WEBHOOK_URL=https://hooks.slack.com/services/.......
 
 The values of each key are irrelevant for the test setup. For production they need to be kept secret. The email enviromental variables may be left out if emailing functionality is not required.
 
 Now start the services with:
 
 ```
-$ bin/boot.sh
+sudo bin/boot.sh
```
[the remark this answers is not in the document]
Comment (DanielMurphy22) [reply README.md:53]: Ahh... I am that user!

File: web/services/management/commands/removeoldreports.py (hunk 1 above)
Comment (DanielMurphy22) [reply web/services/management/commands/removeoldreports.py]: I like this change! I have had to use `no_refs = UserDetails.objects.annotate(num_refs=Count('errorreports')).filter(num_refs=0)`
('user' to 'errorreports' and no `__eq`) for it to work!

Comment (DavidFair) [reply web/services/management/commands/removeoldreports.py]: Ahh I had the join backwards - thanks for working through my slightly mislead guidance

**Comment by DavidFair** [comment]
Thanks for doing that, code looks good - I need to test this locally and make a backup of the DB before deploying this. But at least code-wise this looks done :)

File: web/services/management/commands/removeoldreports.py (hunk 1 above)
Comment (DanielMurphy22) [reply web/services/management/commands/removeoldreports.py]: Not saying my code doesn't need a once over (a bit freaked out looking at python after 3 weeks of cpp learning), but your comments were enough for me to figure it out!

**Comment by DavidFair** [comment]
@DanielMurphy22 You have some CI warnings to resolve though by the looks of it....

**Review by DavidFair — APPROVED** [review]
Tested and runs correctly locally 
I've included the backup / restore scripts I've tested locally too, just to make it easier to find them in a "panic" situation

**Pull request merged**